File: backend/src/ai_candidate_screening_outreach/schemas/requirements.py
"""Versioned Requirements Profile — the recruiter's structured answers at
campaign creation. Merged with (and overriding) the JD extraction in the
pipeline (Phase 4).

Every filterable criterion carries a `mode`:
  - "off"          not evaluated
  - "preference"   affects scoring, never rejects
  - "hard_filter"  can reject, with PASS/FAIL/UNKNOWN semantics —
                   FAIL only on explicit contradicting resume evidence;
                   UNKNOWN routes to human review, never auto-reject.
"""

from datetime import date
from enum import Enum
from typing import Literal, Optional

from pydantic import BaseModel, Field, model_validator


class Mode(str, Enum):
    off = "off"
    preference = "preference"
    hard_filter = "hard_filter"


class SkillRequirement(BaseModel):
    skill: str = Field(min_length=1, max_length=100)
    min_years: Optional[float] = Field(default=None, ge=0, le=50)


class CustomWeights(BaseModel):
    """Scoring bucket caps; must sum to 100."""

    required_skills: int = Field(default=40, ge=0, le=100)
    must_haves: int = Field(default=25, ge=0, le=100)
    experience: int = Field(default=15, ge=0, le=100)
    education: int = Field(default=10, ge=0, le=100)
    preferred_skills: int = Field(default=10, ge=0, le=100)

    @model_validator(mode="after")
    def _sums_to_100(self):
        total = (
            self.required_skills
            + self.must_haves
            + self.experience
            + self.education
            + self.preferred_skills
        )
        if total != 100:
            raise ValueError(f"Weights must sum to 100 (got {total})")
        return self


WEIGHT_PRESETS: dict[str, CustomWeights] = {
    "balanced": CustomWeights(),
    "skills_first": CustomWeights(
        required_skills=50, must_haves=20, experience=10, education=5, preferred_skills=15
    ),
    "experience_first": CustomWeights(
        required_skills=30, must_haves=20, experience=30, education=10, preferred_skills=10
    ),
}


class RequirementsProfileV1(BaseModel):
    version: Literal[1] = 1

    # ---- 1. Role & context ----
    seniority: Optional[Literal["junior", "mid", "senior", "lead", "manager"]] = None
    openings: Optional[int] = Field(default=None, ge=1, le=500)
    target_join_date: Optional[date] = None
    role_type: Literal["ic", "manager", "either"] = "either"
    industries: list[str] = []
    industries_mode: Mode = Mode.off

    # ---- 2. Location & work mode ----
    work_mode: Optional[Literal["onsite", "hybrid", "remote"]] = None
    hybrid_days_per_week: Optional[int] = Field(default=None, ge=1, le=6)
    office_location: Optional[str] = Field(default=None, max_length=200)
    commute_rule: Optional[Literal["same_city", "metro_area", "radius_km"]] = None
    commute_radius_km: Optional[int] = Field(default=None, ge=1, le=500)
    relocation_acceptable: bool = True
    relocation_assistance: bool = False
    remote_scope: Optional[Literal["in_country", "international"]] = None
    timezone_overlap_zone: Optional[str] = Field(default=None, max_length=50)
    timezone_overlap_hours: Optional[int] = Field(default=None, ge=1, le=12)
    location_mode: Mode = Mode.hard_filter

    # ---- 3. Work authorization (region-conditional) ----
    us_work_auth_required: Optional[bool] = None
    us_sponsorship: Optional[Literal["none", "transfer_only", "new_ok"]] = None
    us_opt_cpt_ok: Optional[bool] = None
    us_employment_type: Optional[Literal["w2", "c2c", "1099", "any"]] = None
    uk_right_to_work_required: Optional[bool] = None
    uk_sponsor_available: Optional[bool] = None
    work_auth_mode: Mode = Mode.off

    # ---- 4. Experience ----
    min_years_experience: Optional[float] = Field(default=None, ge=0, le=50)
    target_years_min: Optional[float] = Field(default=None, ge=0, le=50)
    target_years_max: Optional[float] = Field(default=None, ge=0, le=60)
    hands_on_requirements: list[str] = []
    hands_on_mode: Mode = Mode.off
    company_stage_pref: Optional[Literal["startup", "mnc", "agency", "enterprise"]] = None
    flag_employment_gaps: bool = False
    experience_mode: Mode = Mode.off

    # ---- 5. Skills & qualifications ----
    must_have_skills: list[SkillRequirement] = []
    must_have_skills_mode: Mode = Mode.preference
    nice_to_have_skills: list[str] = []
    certifications: list[str] = []
    certifications_mode: Mode = Mode.off
    licenses: list[str] = []
    licenses_mode: Mode = Mode.off
    portfolio_required: bool = False  # scoring only
    education_degree_required: bool = False
    education_field: Optional[str] = Field(default=None, max_length=200)
    education_equivalent_ok: bool = True
    education_mode: Mode = Mode.off

    # ---- 6. Compensation (flag-only by design — never auto-rejects) ----
    budget_min: Optional[float] = Field(default=None, ge=0)
    budget_max: Optional[float] = Field(default=None, ge=0)
    budget_currency: Optional[Literal["USD", "GBP", "INR"]] = None
    flag_over_budget: bool = True

    # ---- 7. Availability & logistics ----
    max_notice_days: Optional[int] = Field(default=None, ge=0, le=365)
    immediate_joiners_only: bool = False
    shift: Optional[Literal["day", "night", "rotational", "on_call"]] = None
    contract_type: Optional[
        Literal["permanent", "fixed_term", "contract", "contract_to_hire"]
    ] = None
    travel_percent_max: Optional[int] = Field(default=None, ge=0, le=100)
    availability_mode: Mode = Mode.off

    # ---- 8. Language & communication ----
    english_spoken: Optional[Literal["basic", "professional", "fluent"]] = None
    english_written: Optional[Literal["basic", "professional", "fluent"]] = None
    other_languages: list[str] = []
    language_mode: Mode = Mode.off

    # ---- 9. Pipeline behavior ----
    maybe_band: int = Field(default=10, ge=0, le=30)
    weight_preset: Literal["balanced", "skills_first", "experience_first", "custom"] = (
        "balanced"
    )
    custom_weights: Optional[CustomWeights] = None
    dealbreakers: Optional[str] = Field(default=None, max_length=2000)
    max_shortlist: Optional[int] = Field(default=None, ge=1, le=500)

    @model_validator(mode="after")
    def _validate(self):
        if self.weight_preset == "custom" and self.custom_weights is None:
            raise ValueError("custom_weights is required when weight_preset is 'custom'")
        if (
            self.budget_min is not None
            and self.budget_max is not None
            and self.budget_min > self.budget_max
        ):
            raise ValueError("budget_min cannot exceed budget_max")
        if (
            self.target_years_min is not None
            and self.target_years_max is not None
            and self.target_years_min > self.target_years_max
        ):
            raise ValueError("target_years_min cannot exceed target_years_max")
        return self

    def effective_weights(self) -> CustomWeights:
        if self.weight_preset == "custom" and self.custom_weights:
            return self.custom_weights
        return WEIGHT_PRESETS.get(self.weight_preset, WEIGHT_PRESETS["balanced"])

File: frontend/src/lib/requirements.ts
// Mirrors backend RequirementsProfileV1 (schemas/requirements.py).
// The server re-validates authoritatively; this drives the form.

export type Mode = "off" | "preference" | "hard_filter";

export type SkillRequirement = { skill: string; min_years: number | null };

export type CustomWeights = {
  required_skills: number;
  must_haves: number;
  experience: number;
  education: number;
  preferred_skills: number;
};

export type RequirementsProfile = {
  version: 1;
  // 1. Role & context
  seniority: "junior" | "mid" | "senior" | "lead" | "manager" | null;
  openings: number | null;
  target_join_date: string | null;
  role_type: "ic" | "manager" | "either";
  industries: string[];
  industries_mode: Mode;
  // 2. Location & work mode
  work_mode: "onsite" | "hybrid" | "remote" | null;
  hybrid_days_per_week: number | null;
  office_location: string | null;
  commute_rule: "same_city" | "metro_area" | "radius_km" | null;
  commute_radius_km: number | null;
  relocation_acceptable: boolean;
  relocation_assistance: boolean;
  remote_scope: "in_country" | "international" | null;
  timezone_overlap_zone: string | null;
  timezone_overlap_hours: number | null;
  location_mode: Mode;
  // 3. Work authorization
  us_work_auth_required: boolean | null;
  us_sponsorship: "none" | "transfer_only" | "new_ok" | null;
  us_opt_cpt_ok: boolean | null;
  us_employment_type: "w2" | "c2c" | "1099" | "any" | null;
  uk_right_to_work_required: boolean | null;
  uk_sponsor_available: boolean | null;
  work_auth_mode: Mode;
  // 4. Experience
  min_years_experience: number | null;
  target_years_min: number | null;
  target_years_max: number | null;
  hands_on_requirements: string[];
  hands_on_mode: Mode;
  company_stage_pref: "startup" | "mnc" | "agency" | "enterprise" | null;
  flag_employment_gaps: boolean;
  experience_mode: Mode;
  // 5. Skills & qualifications
  must_have_skills: SkillRequirement[];
  must_have_skills_mode: Mode;
  nice_to_have_skills: string[];
  certifications: string[];
  certifications_mode: Mode;
  licenses: string[];
  licenses_mode: Mode;
  portfolio_required: boolean;
  education_degree_required: boolean;
  education_field: string | null;
  education_equivalent_ok: boolean;
  education_mode: Mode;
  // 6. Compensation
  budget_min: number | null;
  budget_max: number | null;
  budget_currency: "USD" | "GBP" | "INR" | null;
  flag_over_budget: boolean;
  // 7. Availability
  max_notice_days: number | null;
  immediate_joiners_only: boolean;
  shift: "day" | "night" | "rotational" | "on_call" | null;
  contract_type: "permanent" | "fixed_term" | "contract" | "contract_to_hire" | null;
  travel_percent_max: number | null;
  availability_mode: Mode;
  // 8. Language
  english_spoken: "basic" | "professional" | "fluent" | null;
  english_written: "basic" | "professional" | "fluent" | null;
  other_languages: string[];
  language_mode: Mode;
  // 9. Pipeline behavior
  maybe_band: number;
  weight_preset: "balanced" | "skills_first" | "experience_first" | "custom";
  custom_weights: CustomWeights | null;
  dealbreakers: string | null;
  max_shortlist: number | null;
  // 10. Role context & culture (client form v2)
  role_title: string | null;
  urgency: "standard" | "high" | "critical" | null;
  team_context: string | null;
  culture_text: string | null;
  positive_signals: string[];
  concern_signals: string[];
};

// Mirrors WEIGHT_PRESETS in backend schemas/requirements.py — keep in sync.
export const WEIGHT_PRESETS: Record<
  Exclude<RequirementsProfile["weight_preset"], "custom">,
  CustomWeights
> = {
  balanced: {
    required_skills: 40,
    must_haves: 25,
    experience: 15,
    education: 10,
    preferred_skills: 10,
  },
  skills_first: {
    required_skills: 50,
    must_haves: 20,
    experience: 10,
    education: 5,
    preferred_skills: 15,
  },
  experience_first: {
    required_skills: 30,
    must_haves: 20,
    experience: 30,
    education: 10,
    preferred_skills: 10,
  },
};

// Recruiter-facing names for the five scoring areas, in display order.
export const WEIGHT_LABELS: Record<keyof CustomWeights, string> = {
  required_skills: "Required skills",
  must_haves: "Must-haves",
  experience: "Experience",
  education: "Education",
  preferred_skills: "Nice-to-have skills",
};

export function weightsTotal(w: CustomWeights): number {
  return (
    w.required_skills +
    w.must_haves +
    w.experience +
    w.education +
    w.preferred_skills
  );
}

export function defaultRequirements(): RequirementsProfile {
  return {
    version: 1,
    seniority: null,
    openings: null,
    target_join_date: null,
    role_type: "either",
    industries: [],
    industries_mode: "off",
    work_mode: null,
    hybrid_days_per_week: null,
    office_location: null,
    commute_rule: null,
    commute_radius_km: null,
    relocation_acceptable: true,
    relocation_assistance: false,
    remote_scope: null,
    timezone_overlap_zone: null,
    timezone_overlap_hours: null,
    location_mode: "hard_filter",
    us_work_auth_required: null,
    us_sponsorship: null,
    us_opt_cpt_ok: null,
    us_employment_type: null,
    uk_right_to_work_required: null,
    uk_sponsor_available: null,
    work_auth_mode: "off",
    min_years_experience: null,
    target_years_min: null,
    target_years_max: null,
    hands_on_requirements: [],
    hands_on_mode: "off",
    company_stage_pref: null,
    flag_employment_gaps: false,
    experience_mode: "off",
    must_have_skills: [],
    must_have_skills_mode: "preference",
    nice_to_have_skills: [],
    certifications: [],
    certifications_mode: "off",
    licenses: [],
    licenses_mode: "off",
    portfolio_required: false,
    education_degree_required: false,
    education_field: null,
    education_equivalent_ok: true,
    education_mode: "off",
    budget_min: null,
    budget_max: null,
    budget_currency: null,
    flag_over_budget: true,
    max_notice_days: null,
    immediate_joiners_only: false,
    shift: null,
    contract_type: null,
    travel_percent_max: null,
    availability_mode: "off",
    english_spoken: null,
    english_written: null,
    other_languages: [],
    language_mode: "off",
    maybe_band: 10,
    weight_preset: "balanced",
    custom_weights: null,
    dealbreakers: null,
    max_shortlist: null,
    role_title: null,
    urgency: null,
    team_context: null,
    culture_text: null,
    positive_signals: [],
    concern_signals: [],
  };
}

export type MyCompany = {
  id: number;
  name: string;
  default_region: "US" | "UK" | "IN";
  default_threshold: number;
  office_locations: string[];
  recruiter_signature: string | null;
  tone_notes: string | null;
  data_retention_days: number | null;
};
